File: android/app/src/main/java/gh/group38/smartsocket/data/ChargeSession.kt
package gh.group38.smartsocket.data

/** One completed charge, recorded at the moment power was cut. */
data class ChargeSession(
    val endedAtMillis: Long,
    val peakMa: Int,
    val cutAtMa: Int,
    val durationMs: Long,
) {
    val peakAmps: Float get() = peakMa / 1000f

    /**
     * How far the current fell before the socket cut, as a fraction of peak.
     * The number the firmware's learning is built on, shown back to the user.
     */
    val taperFraction: Float
        get() = if (peakMa <= 0) 0f else (cutAtMa.toFloat() / peakMa).coerceIn(0f, 1f)

    fun encode(): String = "$endedAtMillis,$peakMa,$cutAtMa,$durationMs"

    companion object {
        fun decode(line: String): ChargeSession? {
            val p = line.split(',')
            if (p.size != 4) return null
            return ChargeSession(
                endedAtMillis = p[0].toLongOrNull() ?: return null,
                peakMa = p[1].toIntOrNull() ?: return null,
                cutAtMa = p[2].toIntOrNull() ?: return null,
                durationMs = p[3].toLongOrNull() ?: return null,
            )
        }
    }
}

File: desktop/src/SmartSocket.Core/ChargeSession.cs
using System.Globalization;

namespace SmartSocket.Core;

/// <summary>One completed charge, recorded at the moment power was cut.</summary>
public sealed record ChargeSession(
    long EndedAtMillis,
    int PeakMa,
    int CutAtMa,
    long DurationMs)
{
    public float PeakAmps => PeakMa / 1000f;

    /// <summary>
    /// How far the current fell before the socket cut, as a fraction of peak.
    /// The number the firmware's learning is built on, shown back to the user.
    /// </summary>
    public float TaperFraction =>
        PeakMa <= 0 ? 0f : Math.Clamp((float)CutAtMa / PeakMa, 0f, 1f);

    public string Encode() =>
        string.Create(
            CultureInfo.InvariantCulture,
            $"{EndedAtMillis},{PeakMa},{CutAtMa},{DurationMs}");

    public static ChargeSession? Decode(string? line)
    {
        if (line is null) return null;

        var p = line.Split(',');
        if (p.Length != 4) return null;

        if (!long.TryParse(p[0], NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var endedAt)) return null;
        if (!int.TryParse(p[1], NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var peak)) return null;
        if (!int.TryParse(p[2], NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var cutAt)) return null;
        if (!long.TryParse(p[3], NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var duration)) return null;

        return new ChargeSession(endedAt, peak, cutAt, duration);
    }
}
